Proposal for new state: SubFlow State

tsurdilo opened the pull request:
State Description (also included in spec.md with this change):
********
It is often the case that you want to group your workflows into small, **reusable** logical units that perform certain needed functionality.
Even tho you can use the Event state to call an externally deployed services (via function), at times
there is a need to include/nest another serverless workflow (from classpath/local file system etc). In that case you would use the SubFlow State.
It also allows users to model their workflows with reusability and logical grouping in mind.

This state allows you to include/nest a uniquely identified sub-workflow and start its execution. 
Another use of this state is within [branches](#parallel-state-branch) of the [Parallel State](#Parallel-State). Instead of having to define all states
in each branch, you could separate the branch states into individual sub-workflows and call the SubFlow state
as a single state in each.

Sub-workflows must have a defined start and end states. 
The wait-for-completion property defines if the SubFlow state should wait until execution of the sub-workflow
is completed or not. 

Each sub-workflow receives a copy of the SubFlow state's input data.
If wait-for-completion property is set to true, sub-workflows have the ability to edit the parent's workflow data.
If this property is sete to false, data access to parent's workflow should not be allowed.
********

Similar states used in other "workflow" solutions:
BPMN2: "Call Activity"
Netflix Condutor: "Sub Workflow"

Depends on https://github.com/cncf/wg-serverless/pull/95 (where we add workflow "id" property) and https://github.com/cncf/wg-serverless/pull/86 (where we add the filter definition).

cathyhongzhang commented on workflow/spec/schema/serverless-workflow-schema-01.json:
```diff
@@ -452,6 +460,47 @@
+        }
+      },
+      "required": ["name", "type", "next-state", "workflow-id"]
+    },
```
Great! I was thinking about adding something like this.  Thank you! 
Just wondering if we need workflow-version since the workflow-id already uniquely identifies the workflow (My understanding of the version is to identify a specific workflow spec if it has several evolving updates)

cathyhongzhang replied:
Note sure if "invoke state" is the best name. Something like "subworkflow state" is more self-explanatory, but the naming is too long. Would you like to provide a better name?

tsurdilo replied:
ok will update both.

tsurdilo replied:
updated

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -911,6 +913,86 @@ The elements of the output array need not be of the same type.
+| end |If this state and end state | boolean | no |
+| wait-for-completion |If workflow execution must wait for local workflow to finish before continuing | boolean | yes |
+| workflow-id |Local workflow unique id | boolean | no |
+| workflow-version |Local workflow version | boolean | no |
```
ditto

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -911,6 +913,86 @@ The elements of the output array need not be of the same type.
+            "description": "Local workflow unique id."
+        },
+        "workflow-version": {
+            "type": "string",
```
ditto

tsurdilo replied:
updated

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -911,6 +913,86 @@ The elements of the output array need not be of the same type.
+| end |If this state and end state | boolean | no |
+| wait-for-completion |If workflow execution must wait for local workflow to finish before continuing | boolean | yes |
+| workflow-id |Local workflow unique id | boolean | no |
+| workflow-version |Local workflow version | boolean | no |
```
ditto

tsurdilo replied:
updated

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -911,6 +913,86 @@ The elements of the output array need not be of the same type.
+</details>
+
+It is often the case that you want to group your workflows into small, **reusable** logical units that perform certain needed functionality.
+Even tho you can use the Event state to call an externally deployed services (via function), at times
```
change tho to though

tsurdilo replied:
updated

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -911,6 +913,86 @@ The elements of the output array need not be of the same type.
+is completed or not. 
+
+Each invoked workflow receives a copy of the Invoke state's input data.
+Included workflows executed are not allowed to edit the parent workflows' data.
```
why do we have this restriction? There might be cases in which the purpose is to have the sub-workflow modify the parent workflow's data.

tsurdilo replied:
Could we add a flag like "independent": true/false so users can choose? If true then the subflow would maintain its own json context object. If false it would have access to the parent workflow context. wdyt?

cathyhongzhang replied:
If wait-for-completion is false, the parent workflow will not wait for the subflow result. In this case we should not allow subflow to modify the main flow's data since we do not want timing to decide the data on which the main workflow operates. If wait-for-completion is true, then we should not restrict the subflow to modify the data. What do you think?

tsurdilo replied:
ok. will update.

tsurdilo replied:
updated

tsurdilo commented:
@cathyhongzhang changed state name to "SubFlow", removed workflow-version property. Updated the information passing info (depending on wait-for-completion) property.

tsurdilo commented:
rebased and updated props to use camelcase.

tsurdilo commented:
rebased again

ruromero commented on workflow/spec/schema/serverless-workflow-schema-01.json:
```diff
@@ -504,6 +512,43 @@
+      "type": "object",
+      "description": "Defines a sub-workflow to be executed",
+      "properties": {
+        "name": {
```
I think you're missing the new `id` property

tsurdilo replied:
updated

ruromero commented on workflow/spec/schema/serverless-workflow-schema-01.json:
```diff
@@ -504,6 +512,43 @@
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "Unique name of the state"
```
If the `id` property is added, this is no longer _Unique_

tsurdilo replied:
updated

ruromero commented on workflow/spec/spec.md:
```diff
@@ -1006,6 +1008,81 @@ The elements of the output array need not be of the same type.
+
+It is often the case that you want to group your workflows into small, **reusable** logical units that perform certain needed functionality.
+Even tho you can use the Event state to call an externally deployed services (via function), at times
+there is a need to include/nest another serverless workflow (from classpath/local file system etc). In that case you would use the SubFlow State.
```
I would add that the concept of local workflow may vary depending on the implementation. It can also be a Kubernetes namespace, for example.

tsurdilo replied:
good point, added "depending on the implementation logic" to note that in doc. wdyt?
